Azure offloader

david-streamlio opened the pull request:
### Motivation

Added Azure offloader for Pulsar Tiered Storage

### Modifications

Refactored the tiered storage configuration tree to make it easier to add other JCloud Storage Providers in the future

### Result

Users will be able to use their Azure Storage accounts for tiered storage of Pulsar data.

jiazhai commented:
@david-streamlio , Thanks for the work.  
There was my typo for "jcloud", of Since this PR brings in some refine work, Would you please also help change "jcloud" to "jclouds" in the file path and code?

sijie commented:
> If possible, Would you please also help change "jcloud" to "jclouds" in the file path and code?

@jiazhai probably we can deal with this rename in a separate PR?

david-streamlio commented:
I am in the process of updating the unit tests, as NONE of them actually used the constructor method that will be used in production.  I thought this would be a say change, but it turns out to be causing issues with the Mocks that wrap the Blobstore we are writing too.

`BlobStore spiedBlobStore = mock(BlobStore.class, delegatesTo(blobStore));`

it appears these mocks don't get invoked, resulting in test failures

sijie commented:
Can we also update the doc page at https://pulsar.incubator.apache.org/docs/en/cookbooks-tiered-storage/?

david-streamlio commented:
@sijie   Yes, I will update that document as well.

david-streamlio commented:
Where is the content for the doc page at https://pulsar.incubator.apache.org/docs/en/cookbooks-tiered-storage/?

sijie commented:
@david-streamlio the content is here : https://github.com/apache/pulsar/blob/master/site2/docs/cookbooks-tiered-storage.md

jiazhai commented on tiered-storage/jcloud/src/test/java/org/apache/bookkeeper/mledger/offload/jcloud/BlobStoreTestBase.java:
```diff
@@ -69,8 +88,9 @@ public void start() throws Exception {
         if (blobStore != null &&
-            (!Boolean.parseBoolean(System.getProperty("testRealGCS", "false")) &&
-             !Boolean.parseBoolean(System.getProperty("testRealGCS", "false")))) {
+            (Boolean.parseBoolean(System.getProperty("testRealAWS", "false")) ||
```
This change seems not right, for real test, we should not call delete container.

jiazhai commented:
retest this please

jiazhai commented on tiered-storage/jcloud/src/main/java/org/apache/bookkeeper/mledger/offload/jclouds/provider/factory/AWSBlobStoreFactory.java:
```diff
@@ -0,0 +1,60 @@
+package org.apache.bookkeeper.mledger.offload.jclouds.provider.factory;
```
license header is missing.

jiazhai commented:
@david-streamlio , Please fix the ut and header issues, Thanks.

jiazhai commented on tiered-storage/jcloud/src/test/java/org/apache/bookkeeper/mledger/offload/jcloud/BlobStoreTestBase.java:
```diff
@@ -88,9 +88,9 @@ public void start() throws Exception {
-            (Boolean.parseBoolean(System.getProperty("testRealAWS", "false")) ||
-             Boolean.parseBoolean(System.getProperty("testRealGCS", "false")) || 
-             Boolean.parseBoolean(System.getProperty("testRealAzure", "false")))) {
+            (!Boolean.parseBoolean(System.getProperty("testRealAWS", "false")) ||
```
"||" should be "&&" ?

jiazhai commented:
rerun java8 tests

jiazhai commented:
retest this please

david-streamlio commented:
What is the expected behavior? So I can confirm the tests are correct.

jiazhai commented:
rerun java8 tests

jiazhai commented:
rerun java8 tests

jiazhai commented:
rerun java8 tests

david-streamlio commented:
I am open to suggestions on how to split these up, but these changes are tightly coupled. I could remove the Glacier offloader, but that would be just one class.

The refactor moved all of the logic for getting the credentials out of a single class with multiple if/the/else if/else if constructs and into a factory.  

Then I refactored the S3, and GCS to use the new factory methods, and added Azure.

david-streamlio commented:
Following up on @ivankelly 's comment. I can submit a PR for the refactoring as it pertains to the AWS S3 offloader first, and then submit a different one for the Azure offloader once the first one is approved.  Will that be the best path forward @sijie ?

sijie commented:
@david-streamlio let's get #2865 merged first. then you can rebase this pull request on #2865

sijie commented:
Close this pull request as Azure Offloader has been merged.
